routes/index.js

```diff
@@ -1,55 +1,406 @@
+    if (req.isAuthenticated()) {
+        User.findOne({
+            _id: req.user._id
+        }, function(err, user) {
+            if (err) return handleError(err)
+            user.library.push(id)
+            user.save(function(err) {
+                if (err) return handleError(err)
+                res.send({
+                    inLibrary: user.library
+                })
+            })
+        })
+    } else {
+        res.send({
+            url: '/user/login'
+        })
+    }
+})
+
+router.delete('/', function(req, res) {
+    var id = req.body.id
+    User.findOne({
+        _id: req.user._id
+    }, function(err, user) {
         if (err) return handleError(err)
-        if (req.isAuthenticated()) {
-            User.findOne({
-                _id: req.user._id
-            }, function(err, user) {
+        var index = user.library.indexOf(id)
+        if (index > -1) {
+            user.library.splice(index, 1)
+        }
+        user.save(function(err) {
+            if (err) return handleError(err)
+            console.log('delete success')
+            res.send({
+                inLibrary: user.library
+            })
+        })
+    })
+})
+
+router.post('/filter', function(req, res) {
+        var langShown = req.body.langShown.toLowerCase()
+        var langFilter = req.body.langFilter.map((lf) => lf.toLowerCase())
+        var songs2d = []
+        var temp = []
+        async.waterfall([
+                function(done) {
+                    Song.find({
+                        source: {
+                            $exists: false
+                        }
+                    }, function(err, songs) {
+                        done(err, songs)
+                    })
+                },
+                function(songs, done) {
+                    var cbs = songs.map((s, i, arr) => {
+                        return function(done) {
+                            temp = []
+                            temp.push(s)
+                            Song.find({
+                                source: s._id
+                            }, (err, translations) => {
+                                translations.forEach((t) => {
+                                    temp.push(t)
+                                })
+                                songs2d.push(temp)
+                                if (i === arr.length - 1) {
+                                    done(null, songs2d)
+                                } else {
+                                    done()
+                                }
+                            })
+                        }
+                    });
+                    async.waterfall(cbs, function(err, songs2d) {
+                        done(null, songs2d)
+                    });
+                },
+                function(songs2d, done) {
+                    songs2d = songs2d.filter((songs) => {
+                        var langArray = (songs.map((song) => song.lang))
```

Comment:
You can use lodash here.
I think something like _.pick(songs, 'lang') could do it. You're code is fine, just teaching you an alternative that could be useful.